src/spesh/disp.c

```diff
@@ -1057,6 +1110,217 @@ static int translate_dispatch_program(MVMThreadContext *tc, MVMSpeshGraph *g,
                         temporaries[target_idx++] = args[source_idx++];
                 }
                 break;
+            case MVMDispOpcodeResultForeignCode: {
+                /* Determine the op we'll specialize to. */
+                MVMOpInfo const *base_op;
+                switch (ins->info->opcode) {
+                    case MVM_OP_dispatch_v:
+                        base_op = MVM_op_get_op(MVM_OP_sp_runnativecall_v);
+                        break;
+                    case MVM_OP_dispatch_o:
+                        base_op = MVM_op_get_op(MVM_OP_sp_runnativecall_o);
+                        break;
+                    case MVM_OP_dispatch_i:
+                        base_op = MVM_op_get_op(MVM_OP_sp_runnativecall_i);
+                        break;
+                    case MVM_OP_dispatch_n:
+                        base_op = MVM_op_get_op(MVM_OP_sp_runnativecall_n);
+                        break;
+                    case MVM_OP_dispatch_s:
+                        base_op = MVM_op_get_op(MVM_OP_sp_runnativecall_s);
+                        break;
+                    default:
+                        MVM_oops(tc, "Unexpected dispatch op when translating bytecode result");
+                }
+
+
+                int box_return_value = 0;
+                int has_return_value = (ins->info->opcode != MVM_OP_dispatch_v);
+                if (ins->info->opcode == MVM_OP_dispatch_o) {
+                    /* Let's see if we can turn it into a more JITable instruction */
+                    MVMSpeshFacts *object_facts = MVM_spesh_get_facts(tc, g, temporaries[op->res_code.temp_invokee]);
+                    MVMNativeCallBody *body = MVM_nativecall_get_nc_body(tc, object_facts->value.o);
+                    if (object_facts->flags & MVM_SPESH_FACT_KNOWN_VALUE) {
```

Comment:
Also here we use `value.o` a bit early, though since I noticed it already, I wonder if we shouldn't check this property (and get the invokee facts) earlier.